src/BenchmarkDotNet.Diagnostics.Windows/NativeMemoryDiagnoser.cs

```diff
@@ -0,0 +1,257 @@
+
+                    cumMetric += data.NewAllocSize;
+                };
+                heapParser.HeapTraceDestroy += delegate (HeapTraceData data)
+                {
+                    if (!start)
+                    {
+                        return;
+                    }
+
+                    // Heap is dieing, kill all objects in it.   
+                    var allocs = lastHeapAllocs;
+                    if (data.HeapHandle != lastHeapHandle)
+                    {
+                        allocs = GetHeap(data.HeapHandle, heaps, ref lastHeapAllocs, ref lastHeapHandle);
+                    }
+
+                    foreach (var alloc in allocs.Values)
+                    {
+                        // TODO this is a clone of the free code.  
+                        cumMetric -= alloc;
+                    }
+                };
+
+                eventSource.Process();
+
+                Logger.WriteLine();
+                Logger.WriteLineHeader(LogSeparator);
+                Logger.WriteLineInfo($"{parameters.DisplayInfo}");
+                Logger.WriteLineHeader(LogSeparator);
+
+                result.totalMemory = totalAllocation / totalOperation;
+                result.memoryLeak = cumMetric / totalOperation;
+
+                Logger.WriteLine($"Total allocated native memory: {result.totalMemory:n3}");
+                if (cumMetric != 0)
+                {
+                    Logger.WriteLine($"Total memory leak: {result.memoryLeak:n3}");
+                }
+                var heapInfoList = heaps.Select(h => new { Address = h.Key, h.Value.Count, types = h.Value.Values });
+                foreach (var item in heapInfoList.Where(p => p.Count > 0))
+                {
+                    Logger.WriteLine($"Count of not deallocated object: {item.Count / totalOperation}");
+                }
+            }
+
+            return result;
+        }
+
+        public RunMode GetRunMode(BenchmarkCase benchmarkCase) => etwProfiler.GetRunMode(benchmarkCase);
+
+        public IEnumerable<Metric> ProcessResults(DiagnoserResults results)
+        {
+            if (!etwProfiler.BenchmarkToEtlFile.TryGetValue(results.BenchmarkCase, out var traceFilePath))
+                yield break;
+
+            var result = ParseEtlFile(results.BenchmarkCase);
+
+            yield return new Metric(new NativeMemoryDescriptor(), result.totalMemory);
+            yield return new Metric(new NativeMemoryLeakDescriptor(), result.memoryLeak);
+        } 
+
+        public IEnumerable<ValidationError> Validate(ValidationParameters validationParameters) => etwProfiler.Validate(validationParameters);
+
+        /// <summary>
+        /// Implements a simple one-element cache for find the heap to look in.  
+        /// </summary>
+        private static Dictionary<Address, long> GetHeap(Address heapHandle, Dictionary<Address, Dictionary<Address, long>> heaps, ref Dictionary<Address, long> lastHeapAllocs, ref Address lastHeapHandle)
```

Comment:
nit: the name of the method is misleading, when I saw it I thought that you are using ETW to connect to a live process and get some heap information. Perhaps `CreateHeapCache` would be a better name?